clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -227,9 +234,83 @@ public void changePartitionAssignment(List<TopicPartition> assignments) {
             if (!this.subscribedTopics.contains(tp.topic()))
                 throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic.");
         this.clearAssignment();
-        this.assignedPartitions.addAll(assignments);
+        for (TopicPartition tp: assignments)
+            this.assignedPartitions.put(tp, new TopicPartitionState());
         this.needsPartitionAssignment = false;
     }
 
+    public boolean isAssigned(TopicPartition tp) {
+        return assignedPartitions.containsKey(tp);
+    }
+
+    public boolean isPaused(TopicPartition tp) {
+        return assertAssignment(tp).paused;
+    }
+
+    public boolean isFetchable(TopicPartition tp) {
+        return fetchable.contains(tp);
+    }
+
+    public void pause(TopicPartition tp) {
+        assertAssignment(tp).pause();
+    }
+
+    public void resume(TopicPartition tp) {
+        assertAssignment(tp).resume();
+    }
+
+    private static class TopicPartitionState {
+        private Long consumed;   // offset exposed to the user
+        private Long fetched;    // current fetch position
+        private Long committed;  // last committed position
+
+        private boolean paused;  // whether this partition has been paused by the user
```

Comment:
Could we move the fetchable Set in SubscriptionState to TopicPartitionState? More specifically:

1) change !paused to fetchable.
2) set fetchable to true in SubscriptionState.seek().
3) set fetchable to in TopicPartitionState.awaitReset().
4) add a fetchablePartitions() in SubscriptionState that loop over the assignedPartition map and add to the returned set if fetchable is true.